### PR TITLE
[Feature] Spam Mixin

### PR DESCRIPTION
**Pull request by Ghalko**
## Proposal
A mixin class that handles spam reports wherever a `StoredObject` uses it. This could be used for:
- Comments (used in this PR)
- Nodes
- Files
- Other user contributed stored objects

The ability to add a way to report spam to any user contributed data without repeating basic functionality makes this DRY. But it also leads to improvements needed for comments and how they are displayed.

## Repercussions
This is a potentially unstable use of MODM, though through testing there doesn't seem to be much instability.

### CODE REVIEW

File: api/comments/serializers.py [hunk 1]
```diff
@@ -62,7 +62,7 @@ def get_is_abuse(self, obj):
         user = self.context['request'].user
         if user.is_anonymous():
             return False
-        return user._id in obj.reports
+        return user._id in obj.reports and not obj.reports[user._id]['retracted']
```
Comment (cwisecarver) [inline api/comments/serializers.py]: Is there a chance that `'retracted'` could cause a `KeyError`? Would you rather use `.get()` with a default or wrap this and the one on line 190 in a `try`? If it's just broken beyond repair if there is a `KeyError` then it's probably worth wrapping it in a try and responding to the user in a more gentle way than a 500.

File: website/project/model.py [hunk 2]
```diff
@@ -155,19 +155,119 @@ class MetaData(GuidStoredObject):
...
+        if self.spam_status == self.UNKNOWN:
+            self.spam_status = self.FLAGGED
+        if save:
+            self.save()
+
+    def remove_flag(self, save=False):
+        if self.spam_status != self.FLAGGED:
+            return
+        for user, report in self.reports.iteritems():
+            if not report['retracted']:
```
Comment (cwisecarver) [inline website/project/model.py]: Again, any chance of `KeyError`?

File: api/comments/serializers.py (hunk 1 above)
Comment (Ghalko) [inline api/comments/serializers.py]: replaced with gets on 65 and 190. The user._id will catch it before a `get` has a chance at a `KeyError`

File: website/project/model.py (hunk 2 above)
Comment (Ghalko) [inline website/project/model.py]: Replaced as above just one level of abstraction further in.

**Comment by Ghalko** [comment]
@cwisecarver Double check?

**Comment by cwisecarver** [comment]
@Ghalko Those three fixes look good. @sloria Zak had talked to Michael about the deployment steps for adding the send grid config to django's settings files. I know default.py is switching the regular SMTP service out for the django mock service so worst case it will print all emails to the console until it's sorted out in the ansible configs.

**Comment by Ghalko** [comment]
For reference ^^^ #4860

File: website/project/model.py [hunk 3]
```diff
@@ -155,19 +155,119 @@ class MetaData(GuidStoredObject):
...
+
+    # Reports is a list of reports
+    # Each report is a dictionary including:
+    #  - user: Reporting user
+    #  - date: date reported
+    #  - retracted: if a report has been retracted
+    #  - category: What type of spam does the reporter believe this is
+    #  - message: Comment on the comment
+    reports = fields.DictionaryField(
+        default={}, validate=validate_reports
```
Comment (sloria) [inline website/project/model.py]: For the same reasons you should not pass mutable values as default parameter values to function, you should not pass a mutable value (e.g. `{}`) as a `Field's` `default`. See http://effbot.org/zone/default-values.htm . You can pass a callable, e.g. `dict`, instead.

File: website/project/model.py [hunk 4]
```diff
@@ -155,19 +155,119 @@ class MetaData(GuidStoredObject):
...
+    )
+
+    def flag_spam(self, save=False):
```
Comment (sloria) [inline website/project/model.py]: Please add unit tests for these new methods. You an add them to the `Comment` model's unit tests.

File: website/project/model.py [hunk 5]
```diff
@@ -155,19 +155,119 @@ class MetaData(GuidStoredObject):
...
+        """Report object is spam or other abuse of OSF
+
+        :param user: User submitting report
+        :param date: Date report submitted
+        :param save: Save changes
+        :param kwargs: Should include category and message
+        :raises ValueError: if user is reporting self
+        """
+        if user == self.user:
+            raise ValueError
```
Comment (sloria) [inline website/project/model.py]: It would be helpful to add a descriptive error message here.

File: website/project/model.py [hunk 6]
```diff
@@ -155,19 +155,119 @@ class MetaData(GuidStoredObject):
...
+            self.spam_status = self.FLAGGED
+        if save:
+            self.save()
+
+    def remove_flag(self, save=False):
+        if self.spam_status != self.FLAGGED:
+            return
+        for user, report in self.reports.iteritems():
+            if not report.get('retracted', True):
+                return
```
Comment (sloria) [inline website/project/model.py]: Why do you return early here?

File: website/project/model.py [hunk 7]
```diff
@@ -155,19 +155,119 @@ class MetaData(GuidStoredObject):
...
+        # If ham and unedited then tell user that they should read it again
+        if self.spam_status == self.UNKNOWN:
+            self.spam_status = self.FLAGGED
+        if save:
+            self.save()
+
+    def remove_flag(self, save=False):
+        if self.spam_status != self.FLAGGED:
+            return
+        for user, report in self.reports.iteritems():
```
Comment (sloria) [inline website/project/model.py]: Minor: since you're not using `user`, you can do

```python
for report in self.reports.values():
   ...
```

File: website/project/model.py [hunk 8]
```diff
@@ -155,19 +155,119 @@ class MetaData(GuidStoredObject):
...
+    def confirm_spam(self, save=False):
+        self.spam_status = self.SPAM
+        if save:
+            self.save()
+
+    @property
+    def is_spam(self):
+        return self.spam_status == self.SPAM
+
+    def report_spam(self, user, save=False, **kwargs):
```
Comment (sloria) [inline website/project/model.py]: Since this method is used for all types of reports, I think the old name, `report_abuse`, is preferable.

**Comment by sloria** [comment]
Pass finished.

File: website/project/model.py (hunk 8 above)
Comment (Ghalko) [inline website/project/model.py]: Sounds good.

File: website/project/model.py (hunk 4 above)
Comment (Ghalko) [inline website/project/model.py]: `TestComment` class was removed in favor of API tests it seems. Searching for comment in test_models.py, where it resided before, comes up empty.

File: website/project/model.py (hunk 5 above)
Comment (Ghalko) [inline website/project/model.py]: Added.

File: website/project/model.py (hunk 6 above)
Comment (Ghalko) [inline website/project/model.py]: If it isn't flagged, don't bother going through the dictionary of reports.

File: website/project/model.py (hunk 7 above)
Comment (Ghalko) [inline website/project/model.py]: Done.

File: website/project/model.py (hunk 8 above)
Comment (Ghalko) [inline website/project/model.py]: Reverted

File: website/project/model.py (hunk 3 above)
Comment (Ghalko) [inline website/project/model.py]: Done.

**Comment by Ghalko** [comment]
@sloria I haven't committed my review response, because I want feedback on what to do with the unittests. I can certainly add them as `TestSpamMixin` class using the comment framework that wasn't completely removed from the factory imports.

File: api_tests/comments/views/test_comment_report_list.py [hunk 9]
```diff
@@ -191,6 +207,8 @@ def test_private_node_logged_in_contributor_can_report_comment(self):
         comment = CommentFactory(node=self.private_project, user=self.contributor)
         url = '/{}comments/{}/reports/'.format(API_BASE, comment._id)
 
+        get_res = self.app.get(url, auth=self.user.auth)
```
Comment (sloria) [inline api_tests/comments/views/test_comment_report_list.py]: Why was this added?

File: tests/test_spam_mixin.py [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+
+
+class TestSpamMixin(OsfTestCase):
+
+    def setUp(self):
+        super(TestSpamMixin, self).setUp()
+        self.comment = CommentFactory()
+        self.auth = Auth(user=self.comment.user)
+
+    @mock.patch('website.project.model.SpamMixin.flag_spam')
```
Comment (sloria) [inline tests/test_spam_mixin.py]: Why do you mock flag_spam?

File: tests/test_spam_mixin.py [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+            date=time,
+            category='spam',
+            text='ads',
+            retracted=False
+        )
+        assert_in(user._id, self.comment.reports)
+        assert_equal(self.comment.reports[user._id], equivalent)
+        assert_equal(len(mock_flag_spam.mock_calls), 1)
+
+    @mock.patch('website.project.model.SpamMixin.flag_spam')
```
Comment (sloria) [inline tests/test_spam_mixin.py]: I don't think the mocking is necessary. Instead, you can assert that spam_status is set to the correct value.

File: website/project/model.py [hunk 12]
```diff
@@ -157,19 +157,121 @@ class MetaData(GuidStoredObject):
...
+        ):
             raise ValidationValueError(
-                'Values must include `category` and `text` keys'
+                ('Values must include `date`, `category`, ',
+                 '`text`, `retracted` keys')
             )
 
 
-class Comment(GuidStoredObject):
+class SpamMixin(StoredObject):
```
Comment (sloria) [inline website/project/model.py]: In order to slim down `model.py`, let's move this mixin to its own module in `website.project.spam`.

**Comment by sloria** [comment]
Pass finished. After the above changes are made, this should be good to merge.

File: api_tests/comments/views/test_comment_report_list.py (hunk 9 above)
Comment (Ghalko) [inline api_tests/comments/views/test_comment_report_list.py]: Debug, removed.

File: tests/test_spam_mixin.py (hunk 10 above)
Comment (Ghalko) [inline tests/test_spam_mixin.py]: Because this is a unit test and the `test_flag_spam` tests that unit. In one case I do have to make it more of an integration test.

File: tests/test_spam_mixin.py (hunk 11 above)
Comment (Ghalko) [inline tests/test_spam_mixin.py]: The status changing functions have their own tests.

File: website/project/model.py (hunk 12 above)
Comment (Ghalko) [inline website/project/model.py]: Moved this to `website.project.spam.model.py` though others seem to follow a different convention: `website.spam.model.py` @sloria opinion?
